File: EncryptoBloc/chaincode/cid-store.js
'use strict';

const { Contract } = require('fabric-contract-api');

class CIDStoreContract extends Contract {

    // Initialize the chaincode
    async InitLedger(ctx) {
        console.info('============= START : Initialize Ledger ===========');
        console.info('============= END : Initialize Ledger ===========');
    }

    // Create composite key for storing records
    createKey(owner, cid) {
        return `cid_${owner}_${cid}`;
    }

    // StoreCID stores a new CID record on the blockchain
    async StoreCID(ctx, cid, fileName) {
        console.info('============= START : Store CID ===========');

        // Get the ID of the submitting client
        const clientID = ctx.clientIdentity.getID();

        const record = {
            cid: cid,
            owner: clientID,
            timestamp: new Date().toISOString(),
            fileName: fileName,
            docType: 'cidRecord' // for query purposes
        };

        // Create composite key
        const key = this.createKey(clientID, cid);

        // Store the record
        await ctx.stub.putState(key, Buffer.from(JSON.stringify(record)));
        console.info('============= END : Store CID ===========');
        return JSON.stringify(record);
    }

    // GetCID retrieves a CID record by its ID
    async GetCID(ctx, cid) {
        console.info('============= START : Get CID ===========');
        
        const clientID = ctx.clientIdentity.getID();
        const key = this.createKey(clientID, cid);
        
        const recordAsBytes = await ctx.stub.getState(key);
        if (!recordAsBytes || recordAsBytes.length === 0) {
            throw new Error(`The CID ${cid} does not exist`);
        }

        console.info('============= END : Get CID ===========');
        return recordAsBytes.toString();
    }

    // GetAllCIDs returns all CIDs stored by the current user
    async GetAllCIDs(ctx) {
        console.info('============= START : Get All CIDs ===========');
        
        const clientID = ctx.clientIdentity.getID();
        
        // Get all records for the owner using composite key
        const startKey = this.createKey(clientID, '');
        const endKey = this.createKey(clientID, '\uffff');

        const iterator = await ctx.stub.getStateByRange(startKey, endKey);
        const records = [];

        while (true) {
            const result = await iterator.next();

            if (result.value && result.value.value.toString()) {
                let record;
                try {
                    record = JSON.parse(result.value.value.toString('utf8'));
                } catch (err) {
                    console.log(err);
                    record = result.value.value.toString('utf8');
                }
                records.push(record);
            }

            if (result.done) {
                await iterator.close();
                console.info('============= END : Get All CIDs ===========');
                return JSON.stringify(records);
            }
        }
    }

    // QueryCIDsByOwner allows querying CIDs by owner using rich queries
    async QueryCIDsByOwner(ctx, owner) {
        console.info('============= START : Query CIDs by Owner ===========');
        
        let queryString = {
            selector: {
                docType: 'cidRecord',
                owner: owner
            }
        };

        let iterator = await ctx.stub.getQueryResult(JSON.stringify(queryString));
        let records = [];

        while (true) {
            const result = await iterator.next();

            if (result.value && result.value.value.toString()) {
                let record;
                try {
                    record = JSON.parse(result.value.value.toString('utf8'));
                    records.push(record);
                } catch (err) {
                    console.log(err);
                }
            }

            if (result.done) {
                await iterator.close();
                console.info('============= END : Query CIDs by Owner ===========');
                return JSON.stringify(records);
            }
        }
    }
}

module.exports = CIDStoreContract; 

File: EncryptoBloc/chaincode/cid-store.go
package main

import (
    "encoding/json"
    "fmt"
    "time"

    "github.com/hyperledger/fabric-contract-api-go/contractapi"
)

// SmartContract provides functions for managing CIDs
type SmartContract struct {
    contractapi.Contract
}

// CIDRecord represents a CID entry with metadata
type CIDRecord struct {
    CID       string    `json:"cid"`
    Owner     string    `json:"owner"`
    Timestamp time.Time `json:"timestamp"`
    FileName  string    `json:"fileName"`
}

// StoreCID stores a new CID record on the blockchain
func (s *SmartContract) StoreCID(ctx contractapi.TransactionContextInterface, cid string, fileName string) error {
    // Get the ID of the submitting client
    owner, err := ctx.GetClientIdentity().GetID()
    if err != nil {
        return fmt.Errorf("failed to get client identity: %v", err)
    }

    record := CIDRecord{
        CID:       cid,
        Owner:     owner,
        Timestamp: time.Now(),
        FileName:  fileName,
    }

    recordJSON, err := json.Marshal(record)
    if err != nil {
        return err
    }

    // Store the CID record using composite key of owner + cid
    key := createKey(owner, cid)
    return ctx.GetStub().PutState(key, recordJSON)
}

// GetCID retrieves a CID record by its ID
func (s *SmartContract) GetCID(ctx contractapi.TransactionContextInterface, cid string) (*CIDRecord, error) {
    owner, err := ctx.GetClientIdentity().GetID()
    if err != nil {
        return nil, fmt.Errorf("failed to get client identity: %v", err)
    }

    key := createKey(owner, cid)
    recordJSON, err := ctx.GetStub().GetState(key)
    if err != nil {
        return nil, fmt.Errorf("failed to read from world state: %v", err)
    }
    if recordJSON == nil {
        return nil, fmt.Errorf("the CID %s does not exist", cid)
    }

    var record CIDRecord
    err = json.Unmarshal(recordJSON, &record)
    if err != nil {
        return nil, err
    }

    return &record, nil
}

// GetAllCIDs returns all CIDs stored by the current user
func (s *SmartContract) GetAllCIDs(ctx contractapi.TransactionContextInterface) ([]*CIDRecord, error) {
    owner, err := ctx.GetClientIdentity().GetID()
    if err != nil {
        return nil, fmt.Errorf("failed to get client identity: %v", err)
    }

    // Get all records for the owner using composite key
    startKey := createKey(owner, "")
    endKey := createKey(owner, "\xff")

    resultsIterator, err := ctx.GetStub().GetStateByRange(startKey, endKey)
    if err != nil {
        return nil, err
    }
    defer resultsIterator.Close()

    var records []*CIDRecord
    for resultsIterator.HasNext() {
        queryResponse, err := resultsIterator.Next()
        if err != nil {
            return nil, err
        }

        var record CIDRecord
        err = json.Unmarshal(queryResponse.Value, &record)
        if err != nil {
            return nil, err
        }
        records = append(records, &record)
    }

    return records, nil
}

// Helper function to create composite keys
func createKey(owner string, cid string) string {
    return fmt.Sprintf("cid_%s_%s", owner, cid)
}

func main() {
    chaincode, err := contractapi.NewChaincode(&SmartContract{})
    if err != nil {
        fmt.Printf("Error creating CID store chaincode: %s", err.Error())
        return
    }

    if err := chaincode.Start(); err != nil {
        fmt.Printf("Error starting CID store chaincode: %s", err.Error())
    }
} 